<a href="https://colab.research.google.com/github/maps-05/portfolio/blob/main/Predictivemodeling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Predictive Modeling: Regression and Classification Analysis

Midterm and Finals Exam

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

In [ ]:
# Load dataset
df = pd.read_csv("https://raw.githubusercontent.com/renatomaaliw3/public_files/refs/heads/master/Data%20Sets/linear-m08.csv")
df

,X,Y
0,37.454012,147.924161
1,95.071431,380.273424
2,73.199394,285.206356
3,59.865848,207.992396
4,15.601864,46.638231
...,...,...
495,35.335223,123.055904
496,58.365611,204.490850
497,7.773464,58.090340
498,97.439481,364.676012


In [ ]:
print("Missing values:\n", df.isnull().sum())

Missing values:
 X    0
Y    0
dtype: int64


In [ ]:
x = df['X']
y = df['Y']

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.25, random_state=101)

# Reshape inputs to 2D arrays
X_train = np.array(X_train).reshape(-1, 1)
X_test = np.array(X_test).reshape(-1, 1)
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

In [ ]:
# Slope
slope = model.coef_[0]
print("Slope of the line:", slope)


Slope of the line: 3.5436065351106114


In [ ]:
y_pred = model.predict(X_test)

# RMSE
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print("RMSE on test data:", rmse)


RMSE on test data: 19.947278366555025


In [ ]:
# Predict Y for X = 38.85
x_input = np.array([[38.85]])
y_pred_input = model.predict(x_input)[0]
print("Predicted Y for X = 38.85:", y_pred_input)

Predicted Y for X = 38.85: 146.03333143153571


# FINALS

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, classification_report

In [ ]:
# Load Dataset

df = pd.read_csv('https://raw.githubusercontent.com/renatomaaliw3/public_files/refs/heads/master/Data%20Sets/classification-f08.csv')
df

,A,B,C,Z
0,39.879094,32.870194,27.903927,0
1,18.360727,19.322001,43.150493,1
2,15.949652,20.044108,44.065608,1
3,26.502751,30.394497,29.064790,1
4,17.020317,13.985204,47.046420,1
...,...,...,...,...
495,NaN,NaN,NaN,1
496,NaN,NaN,NaN,1
497,NaN,NaN,NaN,0
498,NaN,NaN,NaN,1


In [ ]:
df_base = df.iloc[:495]

# Calculate conditional means for Z = 0 and Z = 1
means_z0 = df_base[df_base['Z'] == 0].iloc[:, :-1].mean()
means_z1 = df_base[df_base['Z'] == 1].iloc[:, :-1].mean()

# Replace missing values in index 495 to 499 based on Z value
for idx in range(495, 500):
    z_value = df.loc[idx, 'Z']
    if z_value == 0:
        df.loc[idx, df.columns[:-1]] = df.loc[idx, df.columns[:-1]].fillna(means_z0)
    elif z_value == 1:
        df.loc[idx, df.columns[:-1]] = df.loc[idx, df.columns[:-1]].fillna(means_z1)

df

,A,B,C,Z
0,39.879094,32.870194,27.903927,0
1,18.360727,19.322001,43.150493,1
2,15.949652,20.044108,44.065608,1
3,26.502751,30.394497,29.064790,1
4,17.020317,13.985204,47.046420,1
...,...,...,...,...
495,23.219841,28.612393,32.341908,1
496,23.219841,28.612393,32.341908,1
497,38.856589,30.034736,30.908866,0
498,23.219841,28.612393,32.341908,1


In [ ]:
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.svm import SVC
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, classification_report

X = df.iloc[:, :-1]
y = df['Z']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=101)

# Scale the data
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

param_grid = {
    'C': [0.01, 0.05, 0.1, 0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4, 4.5, 5],
    'kernel': ['linear', 'rbf', 'poly']
}

svc = SVC()
grid = GridSearchCV(svc, param_grid)
grid.fit(X_train_scaled, y_train)

best_model = grid.best_estimator_
y_pred = best_model.predict(X_test_scaled)
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)

# Predict for new data [30, 25, 35]
new_input = np.array([[30, 25, 35]])
new_input_scaled = scaler.transform(new_input)
prediction = best_model.predict(new_input_scaled)

# Print results
print("Optimized value of C:", grid.best_params_['C'])
print("Best kernel:", grid.best_params_['kernel'])
print("Accuracy on test set:", accuracy)
print("Classification Report:\n", report)
print("Prediction for [30, 25, 35]:", prediction[0])

Optimized value of C: 5
Best kernel: rbf
Accuracy on test set: 0.968
Classification Report:
               precision    recall  f1-score   support

           0       0.97      0.97      0.97        70
           1       0.96      0.96      0.96        55

    accuracy                           0.97       125
   macro avg       0.97      0.97      0.97       125
weighted avg       0.97      0.97      0.97       125

Prediction for [30, 25, 35]: 1


In [ ]:
X_train_scaled

array([[ 0.88092415,  0.12510127, -0.15756129],
       [-0.85540697, -0.72771298,  0.63889011],
       [-0.23954995,  1.41417553, -1.57702322],
       ...,
       [-1.42742429, -1.02559646,  1.02781017],
       [-1.22951105, -1.6628883 ,  1.40543516],
       [ 0.54996936,  0.96268072, -0.72123307]])

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, classification_report

dt = DecisionTreeClassifier(random_state=42)
dt.fit(X_train, y_train)

y_pred_dt = dt.predict(X_test)
accuracy_dt = accuracy_score(y_test, y_pred_dt)
report_dt = classification_report(y_test, y_pred_dt)

importances = dt.feature_importances_
feature_names = X.columns
if dt.tree_.feature[0] != -2:
    root_node = feature_names[dt.tree_.feature[0]]
else:
    root_node = "Cannot determine root node for this tree."

least_important = feature_names[np.argmin(importances)]

# Prediction for [35, 30, 25]
new_input_dt = np.array([[35, 30, 25]])
prediction_dt = dt.predict(new_input_dt)

print("Accuracy (Decision Tree):", accuracy_dt)
print("Classification Report (Decision Tree):\n", report_dt)
print("Root node feature:", root_node)
print("Least important feature:", least_important)
print("Prediction for new input [35, 30, 25]:", prediction_dt[0])

Accuracy (Decision Tree): 0.928
Classification Report (Decision Tree):
               precision    recall  f1-score   support

           0       0.97      0.90      0.93        70
           1       0.88      0.96      0.92        55

    accuracy                           0.93       125
   macro avg       0.93      0.93      0.93       125
weighted avg       0.93      0.93      0.93       125

Root node feature: A
Least important feature: B
Prediction for new input [35, 30, 25]: 1


In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
from collections import Counter
import numpy as np


param_grid_rf = {
    'n_estimators': [11, 25, 65, 101, 151, 201, 301, 401],
    'max_features': [2, 3]
}

rf = RandomForestClassifier(random_state=42)
grid_rf = GridSearchCV(rf, param_grid_rf, cv=5, n_jobs=-1)
grid_rf.fit(X_train, y_train)

best_rf = grid_rf.best_estimator_
y_pred_rf = best_rf.predict(X_test)
accuracy_rf = accuracy_score(y_test, y_pred_rf)
report_rf = classification_report(y_test, y_pred_rf)

root_nodes = [tree.tree_.feature[0] for tree in best_rf.estimators_]
root_node_counts = Counter([X.columns[i] for i in root_nodes if i != -2])
most_common_root = root_node_counts.most_common(1)[0][0]

importances_rf = best_rf.feature_importances_
least_important_rf = X.columns[np.argmin(importances_rf)]

# Prediction for [35, 25, 30]
new_input_rf = np.array([[35, 25, 30]])
prediction_rf = best_rf.predict(new_input_rf)

print("Accuracy (Random Forest):", accuracy_rf)
print("Classification Report (Random Forest):\n", report_rf)
print("Most common root node feature:", most_common_root)
print("Least important feature:", least_important_rf)
print("Prediction for new input [35, 25, 30]:", prediction_rf[0])

Accuracy (Random Forest): 0.952
Classification Report (Random Forest):
               precision    recall  f1-score   support

           0       0.97      0.94      0.96        70
           1       0.93      0.96      0.95        55

    accuracy                           0.95       125
   macro avg       0.95      0.95      0.95       125
weighted avg       0.95      0.95      0.95       125

Most common root node feature: A
Least important feature: B
Prediction for new input [35, 25, 30]: 0


## Conclusion

This notebook demonstrates two types of predictive modeling tasks: linear regression and classification.

### Linear Regression (Midterm):
- A linear regression model was built to predict 'Y' based on 'X'.
- The calculated slope of the line was approximately **3.54**.
- The model achieved an RMSE of approximately **19.95** on the test data, indicating the average magnitude of the errors.
- For an input 'X' of 38.85, the predicted 'Y' was approximately **146.03**.

### Classification (Finals):
- Missing values in the classification dataset were imputed using conditional means based on the 'Z' variable.
- **Support Vector Machine (SVM):**
    - Optimized using GridSearchCV, the best model utilized a **radial basis function (rbf) kernel** with a 'C' value of **5**.
    - Achieved an accuracy of **96.8%** on the test set.
    - Predicted 'Z' as **1** for the new input `[30, 25, 35]`.
- **Decision Tree Classifier:**
    - Achieved an accuracy of **92.8%** on the test set.
    - Feature 'A' was identified as the **root node** feature, and 'B' as the **least important** feature.
    - Predicted 'Z' as **1** for the new input `[35, 30, 25]`.
- **Random Forest Classifier:**
    - Optimized using GridSearchCV, achieving an accuracy of **95.2%** on the test set.
    - The most common root node feature across the ensemble was 'A', and 'B' was the **least important** feature.
    - Predicted 'Z' as **0** for the new input `[35, 25, 30]`.

Overall, the SVM model demonstrated the highest accuracy among the classification models for this dataset.